File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    public int myAtoi(String s) {
        s=s.trim();
        if(s.length()==0){
            return 0;
        }
        
        int sign=1;
        int i=0;
        if(s.charAt(0)=='-'){
            sign=-1;
            i=1;
        }else if(s.charAt(0)=='+'){
            sign=1;
            i=1;
        }
        
    return (int)solve(s,i,sign,0L);

    }
    long solve(String s,int i,int sign,long n){
        if(i>=s.length()||!Character.isDigit(s.charAt(i))){
            return sign*n;
        }

        n=n*10+(s.charAt(i)-'0');
        if(n*sign>=Integer.MAX_VALUE){
            return Integer.MAX_VALUE;
        }
        if(n*sign<=Integer.MIN_VALUE){
            return Integer.MIN_VALUE;
        }
     return    solve(s,i+1,sign,n);
        
    }
}

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        s=s.strip()
        if not s:
            return 0
        n=0
        c=0
        sign=1
        if s[0]=='-' or s[0]=='+':
            if s[0]=='-':
                sign=-1
            c+=1

        while(c<len(s)):
            if s[c]=='0':
                c+=1
            else:
                break
        for i in range(c,len(s)):
            if s[i].isdigit():
                n=n*10+int(s[i])
            else:
                break
        n=n*sign
        MIN=-2**31
        MAX=2**31-1
        if n<MIN:
            return MIN
        if n>MAX:
            return MAX
        return n
